File: Problem2_ContiguousArray.cpp
// Time Complexity : O(N) where N is length of nums vector
// Space Complexity : O(N) where N is length of nums vector
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach


#include<vector>
#include<bits/stdc++.h>
#include<stdlib.h>
#include<cmath>
#include<time.h>
#include<iostream>

using namespace std;       

class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        
        if (nums.empty()) return 0;
        int gmax=0;
        int start =0;
        int end=0;
        
        unordered_map<int,int> map;
        
        map.insert(make_pair(0,-1));
        int rSum=0;
        
        for(int i=0;i<nums.size();i++){
            rSum= nums[i]==0?rSum-1 :rSum+1;
            
            if(map.count(rSum)){
                if(gmax<i-map.at(rSum)){
                    gmax=i-map.at(rSum);
                    end=i;start=map.at(rSum)+1;
                }
            }
            //    gmax= max(gmax,i - map.at(rSum));
            else{
                map.insert(make_pair(rSum,i));
            }
            
        }
        std::cout<<start<<endl<<end<<endl;
        return gmax;      
        
        
        
                 
                   
                  
                   
        
    }
};

int main(){
    Solution a;
    vector<int> nums={0,1,0,0,0,1,1,1,0};
    //int target=2;
    int b = a.findMaxLength(nums);
    // for (int x : b) 
    //     cout << x << " "; 
}

File: Problem2_ContiguousArray.cs
// Time Complexity :O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :

//Approach: Running sum pattern is used to identify the length of the subarrays.
//If '0' -> Decrement running sum by 1 else '1' -> Increment running sum by 1.
//Hashmap is used to store the running sum and the index of its first occurance.
//While iterating through the array, if the running sum for the current index exists
//in the hashmap, then we can calculate the length of the balanced subarray by taking
//the difference between the current index and the first index at which the running sum
//occurred and get the maxlength.

public class Solution {
    public int FindMaxLength(int[] nums) {
        int rSum = 0;
        int maxLength = 0;
        Dictionary<int, int> hashMap = new Dictionary<int, int>();
        hashMap.Add(0, -1); //Dummy Value

        for(int i = 0; i < nums.Length; i++) {
            if(nums[i] == 0) {
                rSum = rSum - 1;
            }
            else {
                rSum = rSum + 1;
            }
            if(hashMap.TryGetValue(rSum, out var y)) {
                int arrLength = i - y;
                if (arrLength > maxLength) {
                    maxLength = arrLength;
                }
            }
            else {
                hashMap.Add(rSum, i);
            }
        }
        return maxLength;
    }
}
